### PR TITLE
Support alignment of embeddings on 16/32/64-byte boundaries

### PR DESCRIPTION
**Pull request by danieldk**
This is the variant with aligned embeddings. No full review necessary yet, I want to try/time this with some other code to see the impact.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: rust
 rust:
-  - 1.32.0
+  - 1.36.0
```
Comment (sebpuetz) [inline .travis.yml]: I'm pretty happy with moving the required version up. Going above `1.34` unlocks `TryFrom` ;)

**Comment by danieldk** [comment]
> I haven't gone through everything yet, but I want to clarify some concepts:
> 
> In the previous iteration, only the beginning of the matrix was aligned. In this iteration, the offset of each embedding into the matrix is aligned aswell, giving zero-copy access for viewable embeddings.
> 
> This works by:
> 
>     1. allocating the backing array on a boundary
> 
>     2. padding after each embedding to the next boundary.
> 
>     3. in order to properly access the embeddings, slicing is required since what `ndarray` treats as a row is not necessarily all part of the embedding, so the actual offset is obtained through `MatrixLayout::cols`.

Completely correct.

**Comment by sebpuetz** [comment]
Great, then I think this looks good.

Although I'm not a fan of the `.slice(s![])` syntax, but I guess we have to deal with what we get from `ndarray` here.

**Comment by danieldk** [comment]
> Great, then I think this looks good.
> 
> Although I'm not a fan of the `.slice(s![])` syntax, but I guess we have to deal with what we get from `ndarray` here.

It seems that most cases are covered now, but I need to add some unit tests that check that embeddings are actually aligned. That would also catch the thing that I missed now ;), namely that the boundary is not respected when subword units are used.

I guess that this could not be supported in mmap'ed arrays. For quantized embeddings we could return aligned embeddings.

One thing I am not completely happy about is that you from the outside (e.g. the enum wrappers), you can't really now if the embeddings are aligned, except for manually checking.

**Comment by danieldk** [comment]
Performance compared to master:

```
group                           align                                  master
-----                           -----                                  ------
ndarray-lookup-allround         1.03   461.0±61.04ns        ? B/sec    1.00   447.3±56.43ns        ? B/sec
ndarray-lookup-known            1.01   234.6±29.38ns        ? B/sec    1.00   233.2±23.22ns        ? B/sec
ndarray-lookup-unknown          1.01     18.3±2.16µs        ? B/sec    1.00     18.1±2.01µs        ? B/sec
ndarray-mmap-lookup-allround    1.03  849.9±102.67ns        ? B/sec    1.00  825.6±101.60ns        ? B/sec
ndarray-mmap-lookup-known       1.01   493.3±32.29ns        ? B/sec    1.00   488.5±32.22ns        ? B/sec
ndarray-mmap-lookup-unknown     1.02     23.0±2.34µs        ? B/sec    1.00     22.6±2.26µs        ? B/sec
opq-lookup-allround             1.00    44.6±15.23µs        ? B/sec    1.00    44.6±15.44µs        ? B/sec
opq-lookup-known                1.00     20.7±0.67µs        ? B/sec    1.01     20.9±0.55µs        ? B/sec
opq-lookup-unknown              1.00  963.1±113.43µs        ? B/sec    1.01  974.7±115.92µs        ? B/sec
opq-mmap-lookup-allround        1.00    45.0±15.84µs        ? B/sec    1.00    45.0±15.42µs        ? B/sec
opq-mmap-lookup-known           1.00     20.8±0.59µs        ? B/sec    1.00     20.9±0.55µs        ? B/sec
opq-mmap-lookup-unknown         1.00  964.3±112.83µs        ? B/sec    1.01  969.7±112.98µs        ? B/sec
```

**Comment by sebpuetz** [comment]
I guess it makes more sense if we keep the discussion in this PR, so I'm repeating my question here:

How come mmap is slower although mmap isn't aligned anyways?

**Comment by danieldk** [comment]
> I guess it makes more sense if we keep the discussion in this PR, so I'm repeating my question here:
> 
> How come mmap is slower although mmap isn't aligned anyways?

The differences are so small, that they are probably not differences. In reality, there is no difference at all, because the default alignment is 16 bytes anyway on newer glibcs. There will be big differences for align32/align64 (I fear slower, due to less locality).

**Comment by sebpuetz** [comment]
So the conclusion is, alignment didn't introduce a performance regression. But it might bring us gains when we actually need a specific alignment?

**Comment by danieldk** [comment]
Alignments on 32 and 64-byte boundaries (which would be necessary for aligned AVX/AVX512).

```
group                           embedding-align32                      master
-----                           -----------------                      ------
ndarray-lookup-allround         1.07   479.9±57.15ns        ? B/sec    1.00   447.3±56.43ns        ? B/sec
ndarray-lookup-known            1.05   245.8±26.04ns        ? B/sec    1.00   233.2±23.22ns        ? B/sec
ndarray-lookup-unknown          1.04     18.7±1.88µs        ? B/sec    1.00     18.1±2.01µs        ? B/sec
ndarray-mmap-lookup-allround    1.01  830.2±102.22ns        ? B/sec    1.00  825.6±101.60ns        ? B/sec
ndarray-mmap-lookup-known       1.00   489.8±32.49ns        ? B/sec    1.00   488.5±32.22ns        ? B/sec
ndarray-mmap-lookup-unknown     1.01     22.9±2.24µs        ? B/sec    1.00     22.6±2.26µs        ? B/sec
opq-lookup-allround             1.00    44.6±15.47µs        ? B/sec    1.00    44.6±15.44µs        ? B/sec
opq-lookup-known                1.00     20.8±0.47µs        ? B/sec    1.00     20.9±0.55µs        ? B/sec
opq-lookup-unknown              1.00  967.8±112.32µs        ? B/sec    1.01  974.7±115.92µs        ? B/sec
opq-mmap-lookup-allround        1.00    44.6±15.24µs        ? B/sec    1.01    45.0±15.42µs        ? B/sec
opq-mmap-lookup-known           1.00     20.8±0.54µs        ? B/sec    1.00     20.9±0.55µs        ? B/sec
opq-mmap-lookup-unknown         1.00  962.2±112.57µs        ? B/sec    1.01  969.7±112.98µs        ? B/sec
```

```
group                           embedding-align64                      master
-----                           -----------------                      ------
ndarray-lookup-allround         1.09   487.1±59.72ns        ? B/sec    1.00   447.3±56.43ns        ? B/sec
ndarray-lookup-known            1.09   254.6±22.76ns        ? B/sec    1.00   233.2±23.22ns        ? B/sec
ndarray-lookup-unknown          1.03     18.6±1.90µs        ? B/sec    1.00     18.1±2.01µs        ? B/sec
ndarray-mmap-lookup-allround    1.01  835.7±107.99ns        ? B/sec    1.00  825.6±101.60ns        ? B/sec
ndarray-mmap-lookup-known       1.03   501.7±34.98ns        ? B/sec    1.00   488.5±32.22ns        ? B/sec
ndarray-mmap-lookup-unknown     1.01     22.8±2.29µs        ? B/sec    1.00     22.6±2.26µs        ? B/sec
opq-lookup-allround             1.00    43.7±15.13µs        ? B/sec    1.02    44.6±15.44µs        ? B/sec
opq-lookup-known                1.00     20.8±0.56µs        ? B/sec    1.01     20.9±0.55µs        ? B/sec
opq-lookup-unknown              1.00  968.4±113.11µs        ? B/sec    1.01  974.7±115.92µs        ? B/sec
opq-mmap-lookup-allround        1.00    45.0±15.47µs        ? B/sec    1.00    45.0±15.42µs        ? B/sec
opq-mmap-lookup-known           1.00     20.7±0.58µs        ? B/sec    1.01     20.9±0.55µs        ? B/sec
opq-mmap-lookup-unknown         1.00  961.4±113.37µs        ? B/sec    1.01  969.7±112.98µs        ? B/sec
```

**Comment by danieldk** [comment]
> So the conclusion is, alignment didn't introduce a performance regression. But it might bring us gains when we actually need a specific alignment?

Yeah. Now that we have the benchmarks in place, I'll implement the unknown word adds in SIMD with aligned loads/stores to see if there is any improvement. It could be though that the summing is simply too cheap compared to the cache misses for look ups.

I am curious whether we'll see an performance improvement with your n-grams implementation, since they are ordered by frequency, it may introduce more locality.

**Comment by sebpuetz** [comment]
> > So the conclusion is, alignment didn't introduce a performance regression. But it might bring us gains when we actually need a specific alignment?
> 
> Yeah. Now that we have the benchmarks in place, I'll implement the unknown word adds in SIMD with aligned loads/stores to see if there is any improvement. It could be though that the summing is simply too cheap compared to the cache misses for look ups.
> 
> I am curious whether we'll see an performance improvement with your n-grams implementation, since they are ordered by frequency, it may introduce more locality.

I don't have compatible models laying around, once changes at `finalfrontier` are in, I'll train something.

**Comment by sebpuetz** [comment]
> I am curious whether we'll see an performance improvement with your n-grams implementation, since they are ordered by frequency, it may introduce more locality.

I ran a bench on the `ngram_min_count=100` embeddings i trained:

~~~
ndarray-ngrams-lookup-allround                                                                             
                        time:   [557.16 ns 563.25 ns 569.25 ns]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe

ndarray-ngrams-lookup-known                                                                            
                        time:   [377.74 ns 380.45 ns 383.05 ns]
Found 19 outliers among 100 measurements (19.00%)
  2 (2.00%) low severe
  8 (8.00%) low mild
  5 (5.00%) high mild
  4 (4.00%) high severe

ndarray-ngrams-lookup-unknown                                                                             
                        time:   [22.914 us 23.134 us 23.348 us]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild

ndarray-mmap-ngrams-lookup-allround                                                                             
                        time:   [544.08 ns 549.52 ns 554.95 ns]
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

ndarray-mmap-ngrams-lookup-known                                                                            
                        time:   [377.05 ns 380.11 ns 383.21 ns]
Found 16 outliers among 100 measurements (16.00%)
  10 (10.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

ndarray-mmap-ngrams-lookup-unknown                                                                             
                        time:   [22.979 us 23.218 us 23.452 us]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
~~~

The benches were done on `master` with modifications to `benches/array.rs`, so no alignments. If you also want to try something out, the file is on `turing-sfb` under `/home/sebastian/embeddings/finalfrontier/twe-ngram_vocab-structgram-15eps-10ctx-5ns-300dims-30tok_min-100ngram_min-0.75zipf.fifu`.

**Comment by danieldk** [comment]
But how are they compared to similar bucket indexing embeddings?

**Comment by sebpuetz** [comment]
| array       |  buckets   | ngrams |
|------------  |-------------|------------|
| allround   |  418 ns    |   559 ns |
| known     | 241 ns      |  407 ns |
| unknown | 16 ns        | 22 ns    |

| mmap       |  buckets   | ngrams |
|------------  |-------------|------------|
| allround   |  421 ns    |   545 ns |
| known     | 236 ns      |  378 ns  |
| unknown | 16 ns        |  23 ns |

Numbers are not 100% comparable. The structgram model has 710k in-vocab words and mincount 30 doesn't match that number on TWE (neither new nor old release).

**Comment by danieldk** [comment]
I will close this for now. The PR can be used for future implementation. But currently, we still have to see whether the touted benefits pan out.

**Pull request closed without merging**